File: tests/integration/discs/test_dir.c
/*
 * test_dir.c - tests directory listing via opendir()/readdir()/closedir(),
 * which drive OSGBPB against the real DFS catalogue.
 *
 * The disc (built by build_test_discs.sh) also carries data files ALPHA and
 * BETA, so the listing should include them (plus the program itself, TDIR).
 *
 * Expected screen output: each catalogue entry name on its own line, then END.
 */
#include <stddef.h>
#include <dirent.h>
#include <conio.h>

int main(void) {
    DIR *d;
    struct dirent *e;

    clrscr();

    d = opendir(".");
    if (d == NULL) {
        cputs("NODIR\r\n");
        return 1;
    }

    while ((e = readdir(d)) != NULL) {
        cputs(e->d_name);
        cputs("\r\n");
    }

    closedir(d);
    cputs("END\r\n");
    return 0;
}

File: tests/integration/scripted/test_dir.py
"""Integration test: directory listing via opendir()/readdir()/closedir().

Runs discs/test_dir.c (built as TDIR) on a disc that also holds data files
ALPHA and BETA. The listing (driven by OSGBPB against the real DFS catalogue)
must include the program and both data files.
"""

from __future__ import annotations

import pytest

import clib_harness as h


def test_directory_listing(bbc_for_mode, mode):
    h.reset_and_boot(bbc_for_mode, "TDIR", mode)
    h.run_disc_program(bbc_for_mode, "TDIR")

    h.assert_screen(bbc_for_mode, "TDIR")
    h.assert_screen(bbc_for_mode, "ALPHA")
    h.assert_screen(bbc_for_mode, "BETA")
    h.assert_screen(bbc_for_mode, "END")
    h.assert_not_screen(bbc_for_mode, "NODIR")
